source/common/common/log_state_utils.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <sstream>
+
+namespace Envoy {
+
+#define LOG_MEMBER(member) ", " #member ": " << (member)
+
+#define LOG_OPTIONAL_MEMBER(member)                                                                \
+  ", " #member ": " << ((member).has_value() ? absl::StrCat((member).value()) : "null")
+
+// Macro assumes local member variables
+// os (ostream)
+// indent_level (int)
+#define LOG_DETAILS(member)                                                                        \
+  do {                                                                                             \
+    os << spaces << #member ": ";                                                                  \
+    if ((member) != nullptr) {                                                                     \
+      os << "\n";                                                                                  \
+      (member)->logState(os, indent_level + 1);                                                    \
```

Comment:
I'm not a huge fan of this style of printing and indenting TBH. I get that you want to make this efficient and have no allocations, but it's leaking through above into a duck typing of a pure interface (the header map), which seems like a pattern best avoided across if it's going to appear in more places.

Reply:
Yeah, to your point I think there is value in having a structured data print type, and I can think of several use cases where it'd be valuable internally.  From (possibly outdated?) priors I believe there is also value in a minimal allocation state drop, which is what I planned on implementing first.  I think for out-process communication and machine-consumption we'll eventually want to tackle the latter, but I don't think we'd want to use that version in the fatal error handler and I'd like the "core dumps are hard" session state to be human readable where possible.

Reply:
This is a fair point. I think it would be useful to document (i.e. provide the citations) on how allocations during fatal handling are a problem. The main one I can imagine is that we're in a soft OOM situation, and allocation would lead to a hard OOM, would be good to have some more intuition on this one.